codex/stores/repostore.nim

```diff
@@ -107,23 +107,59 @@ method getBlock*(self: RepoStore, cid: Cid): Future[?!Block] {.async.} =
   trace "Got block for cid", cid
   return Block.new(cid, data)
 
-proc getBlockExpirationTimestamp(self: RepoStore, ttl: ?Duration): SecondsSince1970 =
-  let duration = ttl |? self.blockTtl
-  self.clock.now() + duration.seconds
+proc getBlockExpirationEntry(
+  self: RepoStore,
+  cid: Cid,
+  ttl: SecondsSince1970): ?!BatchEntry =
+  ## Get an expiration entry for a batch with timestamp
+  ##
+
+  without key =? createBlockExpirationMetadataKey(cid), err:
+    return failure(err)
+
+  return success((key, ttl.toBytes))
 
 proc getBlockExpirationEntry(
   self: RepoStore,
-  batch: var seq[BatchEntry],
   cid: Cid,
   ttl: ?Duration): ?!BatchEntry =
-  ## Get an expiration entry for a batch
+  ## Get an expiration entry for a batch for duration since "now"
   ##
 
-  without key =? createBlockExpirationMetadataKey(cid), err:
+  let duration = ttl |? self.blockTtl
+  self.getBlockExpirationEntry(cid, self.clock.now() + duration.seconds)
+
+method ensureExpiry*(
+    self: RepoStore,
+    cid: Cid,
+    expiry: SecondsSince1970
+): Future[?!void] {.async.} =
+  ## Ensure that block's assosicated expiry is at least given timestamp
+  ## If the current expiry is lower then it is updated to the given one, otherwise it is left intact
+  ##
+
+  logScope:
+    cid = cid
+
+  if expiry <= 0:
+    return failure(newException(ValueError, "Expiry timestamp must be larger then zero"))
+
+  let expiryKey = createBlockExpirationMetadataKey(cid)
+  let currentExpiry = await self.metaDs.get(expiryKey.tryGet)
+
+  if currentExpiry.isOk and expiry < (!currentExpiry).toSecondsSince1970:
```

Comment:
I think that just treating any error as a missing value is wrong. You should check if error is an instance of `DatastoreKeyNotFound`, if so treat it as missing value, otherwise you should propagate the error to the caller, so `return failure(err)`.

Reply:
Good point, thanks! 👍